feat(ext/net): implement Happy Eyeballs for `Deno.connect` and `Deno.connectTls`

magurotuna opened the pull request:
Implements the Happy Eyeballs v2 ([RFC 8305]) for TCP and TLS connections, enabling faster connection establishment on dual-stack networks by racing IPv6 and IPv4 addresses with staggered parallel attempts.

[RFC 8305]: https://datatracker.ietf.org/doc/html/rfc8305

Closes #25661

## Changes

- Add `ext/net/happy_eyeballs.rs` with the core algorithm implementation
- Update `Deno.connect` and `Deno.connectTls` to use Happy Eyeballs when multiple addresses are resolved
- Add two new options to `Deno.ConnectOptions` and `Deno.ConnectTlsOptions`:
  - `autoSelectFamily`: Enable/disable Happy Eyeballs (default: `true`)
  - `autoSelectFamilyAttemptDelay`: Delay between connection attempts in ms (default: `250`)

## How it works

1. DNS resolution returns multiple addresses (potentially both IPv6 and IPv4)
2. Addresses are interleaved: IPv6 first, then IPv4, alternating (per [RFC 8305] Section 4)
3. First connection attempt starts immediately
4. New attempts start at configurable intervals (default: 250ms) while previous attempts continue in parallel
5. First successful connection wins; all other pending attempts are cancelled

This avoids the common problem where a broken IPv6 path causes long delays before falling back to IPv4.

## Default behavior

`autoSelectFamily` defaults to `true`, so Happy Eyeballs is on by default for `Deno.connect` and `Deno.connectTls`. This is a deliberate choice for parity with Node's `net.connect()`, whose `autoSelectFamily` default is also `true`, and it gives the better out-of-the-box experience on dual-stack networks where a broken IPv6 path would otherwise stall every connection. Callers who want the previous single-address behavior can opt out per-connection with `autoSelectFamily: false`.

Note that this changes the default connection-establishment semantics of a stable API: with multiple resolved addresses, which address wins the race (and therefore what `localAddr` reports) can differ from before. Single-address resolutions are unaffected in practice, since Happy Eyeballs is a no-op when only one address is resolved.

magurotuna committed:
x

magurotuna committed:
test

magurotuna committed:
revert

magurotuna committed:
parallel

magurotuna committed:
fix

magurotuna committed:
fix

magurotuna committed:
fmt

goffrie commented on ext/net/happy_eyeballs.rs:
```diff
@@ -0,0 +1,425 @@
+      }
+
+      // Time to start next connection (staggered start per RFC 8305)
+      _ = sleep_until(next_attempt_at), if has_more_addrs => {
```
If `has_pending = false` & `has_more_addrs = true`, we should probably start the next attempt immediately instead of waiting; in particular if IPv6 is broken and fails immediately then we should quickly fall back to IPv4

magurotuna replied:
That's a good optimization! Implemented that in [e90a168](https://github.com/denoland/deno/pull/31726/commits/e90a168db737846292f674dbbe0de6fd37d19e7e)

magurotuna committed:
Merge branch 'main' into deno-connect-happy-eyeball

magurotuna committed:
fix(ext/net): start next Happy Eyeballs attempt immediately when no attempts are pending

magurotuna committed:
use port 0 as a dest port to get deterministic refusing addr

magurotuna committed:
always clean abort handle on TCP connect failures

magurotuna committed:
lint

magurotuna committed:
lint

bartlomieju committed:
Merge branch 'main' into deno-connect-happy-eyeball

kajukitli reviewed (commented):
solid implementation of RFC 8305. a few observations:

**the good:**
- interleaving logic is correct (IPv6 first, alternating)
- staggered parallel attempts with configurable delay
- immediate fallback when no pending attempts (avoids waiting `attempt_delay` after fast failure)
- good test coverage including edge cases (cancellation, timeout fallback, parallel racing)
- clean API matching Node.js's `autoSelectFamily` options

**questions/suggestions:**

1. **cancel_handle not passed to TLS** — in `op_net_connect_tls`, you call `connect_happy_eyeballs(addrs, attempt_delay, None)` with `None` for cancel_handle, but `op_net_connect_tcp_inner` passes the actual cancel handle. is this intentional? TLS connections could be long-running and users might want to abort them.

2. **`get_refusing_addr()` in tests** — using port 0 as a refusing address is clever, but the comment says "Port 0 is not a valid destination port for TCP connects." this is true but the behavior is OS-dependent. on some systems it might error differently. might be worth a note or using a different approach like `192.0.2.1:1` (TEST-NET-1, guaranteed unroutable).

3. **single address path** — when `addrs.len() == 1`, you skip Happy Eyeballs entirely. makes sense for performance, but it means the `attempted_addresses` tracking won't happen. not sure if that matters for debugging/observability.

4. **error aggregation** — `last_error` only keeps the last failure. for debugging multi-address failures, it might be useful to aggregate errors or at least mention how many addresses were tried. low priority though.

5. **missing integration test** — the unit tests are great, but a real integration test with actual IPv4/IPv6 dual-stack would be nice. though I understand that's hard to set up in CI.

overall this looks good to ship. the core algorithm is correct and matches the RFC.

bartlomieju committed:
Merge branch 'main' into deno-connect-happy-eyeball

bartlomieju commented:
Rebased onto latest `main` and resolved the conflicts. Pushed as a merge commit (no force-push).

The non-trivial part of the merge: `main` recently landed a post-resolution deny check (`check_net_resolved`) on `Deno.connect` / `Deno.connectTls`, which prevents bypassing IP-literal deny rules via numeric hostname aliases (e.g. `2130706433` -> `127.0.0.1`). That check collided with the Happy Eyeballs path:

- For `Deno.connect` (`ext/net/ops.rs`) git produced a *textually* clean auto-merge that was actually broken: it called `check_net_resolved(&addr.ip(), addr.port(), ...)` where `addr` is the `IpAddr { hostname, port }` op argument, which has no `.ip()` method (would not compile).
- For `Deno.connectTls` (`ext/net/ops_tls.rs`) the Happy Eyeballs version had dropped the deny check entirely.

Resolved both by running the deny check over **all** resolved candidate addresses before connecting, since Happy Eyeballs may connect to any of them (failing closed if any resolved IP is denied). Also re-added the `use serde::Deserialize;` import that the merge had dropped (needed by `TcpConnectOptions`).

Addressing the earlier review threads:

- **@goffrie** (immediate fallback when no attempt is pending) — already implemented in e90a168; the `!has_pending && has_more_addrs` arm starts the next attempt immediately instead of waiting `attempt_delay`.
- **@fibibot** (`last_error` semantics) — added a comment noting we surface the chronologically-last failure, matching Node's `net.connect()`.
- **@kajukitli** (cancel handle not passed to TLS) — the `op_net_connect_tls` op has no abort-resource parameter at all, so there is no cancel handle to pass; `None` matches the pre-Happy-Eyeballs behavior (TLS connect was never cancellable). Wiring cancellation into TLS would be a separate change.
- **@kajukitli** (port 0 as a refusing address in tests) — kept; it gives a deterministic *fast* connection refusal, which is what the fast-fallback timing assertions rely on. A non-routable TEST-NET address would hang rather than refuse and break those tests.

Local `cargo test -p deno_net happy_eyeballs` is green (17/17). Leaving the rest to CI.

bartlomieju committed:
refactor(ext/net): split nested ? in start_connect for readability

bartlomieju commented on ext/net/happy_eyeballs.rs:
```diff
@@ -0,0 +1,456 @@
+      biased;
+
+      // A connection attempt completed
+      result = pending.next(), if has_pending => {
```
Addressed — there is now a comment where `last_error` is initialized in `connect_happy_eyeballs` explaining the choice: when every attempt fails we surface the error from the last attempt to fail (chronologically), which matches Node `net.connect()` behavior of reporting the final failed attempt.

bartlomieju commented on ext/net/01_net.js:
```diff
@@ -716,6 +716,11 @@ async function connect(args) {
             undefined,
             cancelRid,
+            {
+              autoSelectFamily: args.autoSelectFamily ?? true,
```
Confirmed intentional. `Deno.connect` / `Deno.connectTls` are Deno-namespace APIs, not WebIDL-defined Web platform interfaces, so there is no conformance requirement to reject `null` here. The `?? true` / `?? 250` semantics deliberately treat both `null` and `undefined` as "use the default" while preserving an explicit `false` / `0`, which matches Node `autoSelectFamily` / `autoSelectFamilyAttemptDelay` handling.

bartlomieju commented on ext/net/happy_eyeballs.rs:
```diff
@@ -0,0 +1,456 @@
+  #[test]
+  fn test_interleave_more_v4() {
+    let v6_1 = SocketAddr::V6(SocketAddrV6::new(Ipv6Addr::LOCALHOST, 80, 0, 0));
+    let v4_1 = SocketAddr::V4(SocketAddrV4::new(Ipv4Addr::LOCALHOST, 80));
```
Done in 065882a018 — the nested `?` is now split across separate lines with a comment noting the outer result is cancellation and the inner one is the connect itself.

bartlomieju commented on ext/net/ops.rs:
```diff
@@ -512,19 +544,30 @@ pub async fn op_net_connect_tcp_inner(
-      .check_net_resolved(&addr.ip(), addr.port(), "Deno.connect()")?;
+    let mut state_ = state.borrow_mut();
+    let permissions = state_.borrow_mut::<PermissionsContainer>();
+    for addr in &addrs {
```
This now denies the connect if *any* resolved address is denied, even though the non-HE branch below only connects to `addrs[0]`. Previously the check was on just the single used address, so a hostname resolving to `[allowed, denied]` could connect to the allowed one; now it fails closed. Correct for the HE race, but a behavior change for `autoSelectFamily: false` / single-attempt. Same pattern in `ext/net/ops_tls.rs`. Suggest gating the all-candidates check on the HE branch and checking only the attempted address otherwise. Follow-up, not blocking.

bartlomieju replied:
Addressed in aed16c4854. The all-candidates deny check is now gated on the branch that actually races (`use_happy_eyeballs = auto_select_family && addrs.len() > 1`); the non-racing path checks only `addrs[0]`, the single address it will connect to. So a hostname resolving to `[allowed, denied]` under `--deny-net=<denied>` connects to the allowed address again with `autoSelectFamily: false` or a single resolution, while the HE race still fails closed if any candidate it may attempt is denied. Same fix applied to `op_net_connect_tls`. Numeric-alias deny spec tests (`deny_net_numeric_hostname_subnet`, `tcp_numeric_deny`) still pass.

bartlomieju reviewed (commented):
Flagging one behavior change to address as a follow-up (not blocking this landing).

The post-resolution deny check now runs `check_net_resolved` over **all** resolved candidate addresses and fails closed if any one is denied. For the Happy Eyeballs path that's exactly right (the connect may land on any candidate). But it also applies when HE doesn't actually race them:

- when `autoSelectFamily: false`, and
- in general the op only connects to `addrs[0]` in the non-HE branch,

yet the check still rejects if *any other* resolved address is denied. So a hostname resolving to e.g. `[allowed_v4, denied_v6]` under `--deny-net=<v6>` now fails the whole connect, where previously (check on just the first/used address) it succeeded. That's a previously-working `Deno.connect` / `Deno.connectTls` now erroring.

"Deny wins across all candidates" is a defensible stance, but it's a real, potentially surprising regression for the non-racing paths, and it affects both `op_net_connect_tcp_inner` (`ext/net/ops.rs`) and `op_net_connect_tls` (`ext/net/ops_tls.rs`). Likely fix: scope the all-candidates check to the branch that actually races (HE enabled + `addrs.len() > 1`), and check only the address(es) we'll really attempt otherwise.

Good to land as-is; let's track this and fix it in a follow-up.

bartlomieju committed:
fix(ext/net): scope post-resolution deny check to attempted addresses

The post-resolution deny check ran check_net_resolved over every resolved
candidate and failed closed if any one was denied. That is correct for the
Happy Eyeballs race (the connect may land on any candidate), but the non-racing
path (autoSelectFamily: false, or a single resolved address) only ever connects
to addrs[0]. Checking all candidates there was a behavior change: a hostname
resolving to [allowed, denied] under --deny-net=<denied> now failed the whole
connect where previously it succeeded.

Gate the all-candidates check on the branch that actually races and check only
the address that will be used otherwise, in both op_net_connect_tcp_inner and
op_net_connect_tls.
